File: app/models/customer.py
# app/models/customer.py

from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, Integer, String, Text
from sqlalchemy.orm import relationship

from utils.db import Base


class Customer(Base):
    __tablename__ = "customers"

    id = Column(Integer, primary_key=True, index=True)

    customer_type = Column(String, nullable=False, default="business")

    first_name = Column(String, nullable=True)
    last_name = Column(String, nullable=True)
    business_name = Column(String, nullable=True)

    email = Column(String, nullable=True)
    phone = Column(String, nullable=True)

    address_line_1 = Column(String, nullable=True)
    address_line_2 = Column(String, nullable=True)
    city = Column(String, nullable=True)
    state = Column(String, nullable=True)
    postal_code = Column(String, nullable=True)

    notes = Column(Text, nullable=True)

    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    sales = relationship("Sale", back_populates="customer")

File: front3/src/types/customer.ts
// front3/src/types/customer.ts

import type { Machine } from "./machine";
import type { Warranty } from "./warranty";
import type { ServiceRecord } from "./serviceRecord";

export interface Customer {
  id: number;
  customer_type: string;
  first_name: string | null;
  last_name: string | null;
  business_name: string | null;
  email: string | null;
  phone: string | null;
  address_line_1: string | null;
  address_line_2: string | null;
  city: string | null;
  state: string | null;
  postal_code: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CustomerSaleDetail {
  id: number;
  customer_id: number;
  machine_id: number;

  asking_price: number | null;
  sale_price: number;
  status: string;

  sold_at: string | null;
  delivered_at: string | null;
  notes: string | null;

  created_at: string;
  updated_at: string;

  machine: Machine;
  warranty: Warranty | null;
  service_records: ServiceRecord[];
}

export interface CustomerDetail extends Customer {
  sales: CustomerSaleDetail[];
}